File: 45-jump-game-ii/jump-game-ii.java
class Solution {
    public int jump(int[] nums) {
        int length = nums.length - 1;
        int index = 0;
        int hops = 0;
        while (index < length) {
            int max_reach = 0;
            int next_index = 0;
//            System.out.println("I can go on for:"+nums[index]);
//            max_reach = nums[index];
            if (length<=index+nums[index]){
                hops++;
                break;
            }
            for (int i = nums[index]; i > 0; i--) {
                int reach = nums[index+i] + i;
//                System.out.println("The max_reach is:"+max_reach+" and the reach is"+reach);
                if (max_reach < reach) {
                    max_reach = reach;
                    next_index = index + i;
                }
            }
            hops++;
//            System.out.println("from "+index);
            index = next_index;
//            System.out.println(" to "+index);
        }
        return hops;

    }
}

File: 45-jump-game-ii/jump-game-ii.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        if len(nums) == 1: return 0
        jumps = 1
        pos = 0
        mx,mxi = 0,0
        while pos+nums[pos] < len(nums)-1:
            frame = nums[pos]
            for index,num in enumerate(nums[pos+1:pos+1+frame]):
                if num+(index-len(nums[pos+1:pos+1+frame]))>=mx:
                    mx = num+(index-len(nums[pos+1:pos+1+frame]))
                    mxi = index+1
            pos += mxi
            mx,mxi = 0,0
            jumps += 1

        return jumps